ValidationConfigurationException: Constraint [unique] is not supported

**Description**

I have an application which uses Neo4j and I want to add the GraphQL plugin.

There is a shared constraint which has a unique constraint.

```
grails.gorm.default.constraints = {
    uniqueName nullable: false, blank: false, unique: true
}
```

Also, there are domain classes which use that shared constraint.
```
class Speaker {

    static mapWith = "neo4j"
    static graphql = true

    String name

    static constraints = {
        name shared: 'uniqueName'
    }

}
```

**Problem**

When I try to run the app I have an exception.

```
Caused by: grails.gorm.validation.exceptions.ValidationConfigurationException: Constraint [unique] is not supported for property [name] of class [class gormgraphql.Speaker] with type [class java.lang.String]
        at grails.gorm.validation.DefaultConstrainedProperty.applyConstraint(DefaultConstrainedProperty.groovy:665)
        at org.grails.datastore.gorm.validation.constraints.eval.DefaultConstraintEvaluator.applySharedConstraints(DefaultConstraintEvaluator.java:198)
        at org.grails.datastore.gorm.validation.constraints.eval.DefaultConstraintEvaluator.evaluate(DefaultConstraintEvaluator.java:176)
        at org.grails.datastore.gorm.validation.constraints.eval.DefaultConstraintEvaluator.evaluate(DefaultConstraintEvaluator.java:80)
        at org.grails.datastore.gorm.validation.constraints.eval.DefaultConstraintEvaluator.evaluate(DefaultConstraintEvaluator.java:75)
        at grails.gorm.validation.PersistentEntityValidator.<init>(PersistentEntityValidator.groovy:51)
        at org.grails.datastore.gorm.validation.constraints.registry.DefaultValidatorRegistry.getValidator(DefaultValidatorRegistry.groovy:68)
        at org.grails.datastore.mapping.model.AbstractMappingContext.getEntityValidator(AbstractMappingContext.java:214)
        at org.grails.gorm.graphql.entity.property.impl.PersistentGraphQLProperty.<init>(PersistentGraphQLProperty.groovy:75)
        at org.grails.gorm.graphql.entity.property.manager.DefaultGraphQLDomainPropertyManager$Builder.getProperties(DefaultGraphQLDomainPropertyManager.groovy:125)
        at org.grails.gorm.graphql.entity.property.manager.DefaultGraphQLDomainPropertyManager$Builder.getProperties(DefaultGraphQLDomainPropertyManager.groovy:87)
        at org.grails.gorm.graphql.types.output.AbstractObjectTypeBuilder.build(AbstractObjectTypeBuilder.groovy:75)
        at org.grails.gorm.graphql.types.DefaultGraphQLTypeManager.getQueryType(DefaultGraphQLTypeManager.groovy:215)
        at org.grails.gorm.graphql.Schema.generate(Schema.groovy:257)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springsource.loaded.ri.ReflectiveInterceptor.jlrMethodInvoke(ReflectiveInterceptor.java:1427)
        at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:162)
        ... 36 common frames omitted
```

**Reproduce the error**

To reproduce the error:

- Download the project from the repository: https://github.com/AmaliaMV/graphql-shared-unique
- Run the project

The versions, I am using, are:

```
grailsVersion=3.3.10
gormVersion=6.1.12.RELEASE
grailsNeo4jPluginVersion=6.2.0.BUILD-SNAPSHOT
```

**Aditional notes**
This error does not happen when the hibernate plugin is used. You can check this on 'hiberante' branch.

--- a/examples/grails3-neo4j-hibernate/src/test/groovy/functional/tests/Neo4jWithHibernateSpec.groovy
+++ b/examples/grails3-neo4j-hibernate/src/test/groovy/functional/tests/Neo4jWithHibernateSpec.groovy
@@ -1,0 +1,37 @@
+package functional.tests
+
+
+import grails.plugins.GrailsPlugin
+import grails.plugins.GrailsPluginManager
+import org.grails.datastore.gorm.neo4j.Neo4jGrailsPlugin
+import org.grails.plugins.MockGrailsPluginManager
+import org.grails.testing.GrailsUnitTest
+import spock.lang.Specification
+
+class Neo4jWithHibernateSpec extends Specification implements GrailsUnitTest {
+
+    @Override
+    Set<String> getIncludePlugins() {
+        ["domainClass"]
+    }
+
+    void "test that both grailsDomainClassMappingContext and neo4jMappingContext are same when using Neo4j with Hibernate"() {
+
+        setup:
+        GrailsPluginManager pluginManager = new MockGrailsPluginManager(grailsApplication)
+        GrailsPlugin hibernate = Mock(GrailsPlugin)
+        hibernate.getName() >> "hibernate"
+        pluginManager.registerMockPlugin(hibernate)
+
+        Neo4jGrailsPlugin neo4jGrailsPlugin = new Neo4jGrailsPlugin()
+        neo4jGrailsPlugin.grailsApplication = grailsApplication
+        neo4jGrailsPlugin.applicationContext = applicationContext
+        neo4jGrailsPlugin.pluginManager = pluginManager
+        this.defineBeans(neo4jGrailsPlugin)
+
+        expect:
+        applicationContext.containsBean("grailsDomainClassMappingContext")
+        applicationContext.containsBean("neo4jMappingContext")
+        applicationContext.getBean("grailsDomainClassMappingContext") != applicationContext.getBean("neo4jMappingContext")
+    }
+}

--- a/examples/grails3-neo4j/src/test/groovy/functional/tests/StandaloneNeo4jSpec.groovy
+++ b/examples/grails3-neo4j/src/test/groovy/functional/tests/StandaloneNeo4jSpec.groovy
@@ -1,0 +1,30 @@
+package functional.tests
+
+
+import org.grails.datastore.gorm.neo4j.Neo4jGrailsPlugin
+import org.grails.plugins.MockGrailsPluginManager
+import org.grails.testing.GrailsUnitTest
+import spock.lang.Specification
+
+class StandaloneNeo4jSpec extends Specification implements GrailsUnitTest {
+
+    @Override
+    Set<String> getIncludePlugins() {
+        ["domainClass"]
+    }
+
+    void "test that both grailsDomainClassMappingContext and neo4jMappingContext are same when using standalone Neo4j"() {
+
+        setup:
+        Neo4jGrailsPlugin plugin = new Neo4jGrailsPlugin()
+        plugin.grailsApplication = grailsApplication
+        plugin.applicationContext = applicationContext
+        plugin.setPluginManager(new MockGrailsPluginManager(grailsApplication))
+        this.defineBeans(plugin)
+
+        expect:
+        applicationContext.containsBean("grailsDomainClassMappingContext")
+        applicationContext.containsBean("neo4jMappingContext")
+        applicationContext.getBean("grailsDomainClassMappingContext") == applicationContext.getBean("neo4jMappingContext")
+    }
+}

--- a/grails-plugin/src/main/groovy/grails/neo4j/bootstrap/Neo4jDataStoreSpringInitializer.groovy
+++ b/grails-plugin/src/main/groovy/grails/neo4j/bootstrap/Neo4jDataStoreSpringInitializer.groovy
@@ -17,6 +17,7 @@
 package grails.neo4j.bootstrap
 
 import grails.neo4j.Neo4jEntity
+import grails.spring.BeanBuilder
 import groovy.transform.InheritConstructors
 import org.grails.datastore.gorm.bootstrap.AbstractDatastoreInitializer
 import org.grails.datastore.gorm.bootstrap.support.ServiceRegistryFactoryBean
@@ -28,7 +29,7 @@ import org.grails.datastore.gorm.plugin.support.PersistenceContextInterceptorAgg
 import org.grails.datastore.gorm.support.AbstractDatastorePersistenceContextInterceptor
 import org.grails.datastore.gorm.support.DatastorePersistenceContextInterceptor
 import org.grails.datastore.mapping.core.grailsversion.GrailsVersion
-import org.springframework.beans.factory.BeanFactory
+import org.springframework.beans.factory.groovy.GroovyBeanDefinitionReader
 import org.springframework.beans.factory.support.BeanDefinitionRegistry
 import org.springframework.context.ApplicationEventPublisher
 import org.springframework.context.ConfigurableApplicationContext
@@ -85,7 +86,11 @@ class Neo4jDataStoreSpringInitializer extends AbstractDatastoreInitializer {
             neo4jPersistenceInterceptor(getPersistenceInterceptorClass(), ref("neo4jDatastore"))
             neo4jPersistenceContextInterceptorAggregator(PersistenceContextInterceptorAggregator)
             if (!secondaryDatastore) {
-                springConfig.addAlias "grailsDomainClassMappingContext", "neo4jMappingContext"
+                if (delegate instanceof BeanBuilder) {
+                    springConfig.addAlias "grailsDomainClassMappingContext", "neo4jMappingContext"
+                } else if (delegate instanceof GroovyBeanDefinitionReader) {
+                    registerAlias "neo4jMappingContext", "grailsDomainClassMappingContext"
+                }
             }
 
             String transactionManagerBeanName = TRANSACTION_MANAGER_BEAN

--- a/grails-plugin/src/main/groovy/grails/neo4j/bootstrap/Neo4jDataStoreSpringInitializer.groovy
+++ b/grails-plugin/src/main/groovy/grails/neo4j/bootstrap/Neo4jDataStoreSpringInitializer.groovy
@@ -84,7 +84,9 @@ class Neo4jDataStoreSpringInitializer extends AbstractDatastoreInitializer {
             neo4jDriver(neo4jDatastore:"getBoltDriver")
             neo4jPersistenceInterceptor(getPersistenceInterceptorClass(), ref("neo4jDatastore"))
             neo4jPersistenceContextInterceptorAggregator(PersistenceContextInterceptorAggregator)
-
+            if (!secondaryDatastore) {
+                springConfig.addAlias "grailsDomainClassMappingContext", "neo4jMappingContext"
+            }
 
             String transactionManagerBeanName = TRANSACTION_MANAGER_BEAN
             if (!containsRegisteredBean(delegate, beanDefinitionRegistry, transactionManagerBeanName)) {

--- a/grails-plugin/src/test/groovy/grails/neo4j/bootstrap/Neo4jDataStoreSpringInitializerSpec.groovy
+++ b/grails-plugin/src/test/groovy/grails/neo4j/bootstrap/Neo4jDataStoreSpringInitializerSpec.groovy
@@ -2,6 +2,7 @@ package grails.neo4j.bootstrap
 
 import grails.gorm.annotation.Entity
 import org.grails.datastore.gorm.neo4j.config.Settings
+import org.grails.spring.DefaultRuntimeSpringConfiguration
 import org.springframework.core.env.MapPropertySource
 import org.springframework.core.env.StandardEnvironment
 import spock.lang.Specification

